aws_cloudwatch_logs to support `json/emf`

### A note for the community

<!-- Please keep this note for the community -->
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment
<!-- Thank you for keeping this note for the community -->


### Use Cases

For now it's possible to send logs in EMF format to Amazon Cloudwatch using the `aws_cloudwatch_logs` sink. However, without the ability to configure the request header to `x-amzn-logs-format: json/emf`, the CloudWatch service can not correctly parse the EMF log. Can we expose this?

### Attempted Solutions

```
sinks:
  emfLogSink:
    type: aws_cloudwatch_logs
    inputs:
      - emfLogSrc
    region: ${AWS_REGION}
    create_missing_group: false
    create_missing_stream: true
    group_name: ${EMF_LOG_GROUP_NAME}
    stream_name: ${TEST_RACK_NAME}
    buffer:
      type: memory 
      max_events: 500
      when_full: drop_newest
    batch:
      timeout_secs: 5
      max_events: 100
    encoding:
      codec: text
```

### Proposal

I think having a boolean flag to turn on/off adding the `x-amzn-logs-format: json/emf` header would be sufficient.

### References

https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/CloudWatch_Embedded_Metric_Format_Specification.html#CloudWatch_Embedded_Metric_Format_Specification_PutLogEvents

### Version

0.21.2

Hey @jszwedko I'm a bit stuck after resolving all but the last compilation error below:

```
$ make test
cargo nextest run --workspace --no-fail-fast --no-default-features --features "default" 
   Compiling vector v0.22.0 (/home/ec2-user/vector)
error[E0759]: `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
   --> src/sinks/aws_cloudwatch_logs/request.rs:231:29
    |
223 |           &self,
    |           ----- this data with an anonymous lifetime `'_`...
...
231 |           Box::pin(async move {
    |  _____________________________^
232 | |             // #12760 this is a relatively convoluted way of changing the headers of a request
233 | |             // about to be sent. https://github.com/awslabs/aws-sdk-rust/issues/537 should
234 | |             // eventually make this better.
...   |
253 | |             client.call(Operation::from_parts(Request::from_parts(body, props), parts)).await
254 | |         })
    | |_________^ ...is captured here, requiring it to live as long as `'static`

For more information about this error, try `rustc --explain E0759`.
error: could not compile `vector` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed
error: command /home/ec2-user/.rustup/toolchains/1.58.1-x86_64-unknown-linux-gnu/bin/cargo test --no-run --message-format json-render-diagnostics --worksp
ace --features default --no-default-features exited with code 101
make: *** [test] Error 101
```

The change can be seen at https://github.com/vectordotdev/vector/pull/12866/files#diff-04a3fbd5514b4e1f92e3a1148d001286eea9aff5479554f7e42921e4e3719bb2R231

Could you provide a bit guidance?

--- a/src/aws/mod.rs
+++ b/src/aws/mod.rs
@@ -82,6 +82,40 @@ pub trait ClientBuilder {
     fn build(client: aws_smithy_client::Client, config: &aws_types::SdkConfig) -> Self::Client;
 }
 
+pub async fn create_smithy_client<T: ClientBuilder>(
+    region: Region, 
+    proxy: &ProxyConfig,
+    tls_options: &Option<TlsConfig>,
+    is_sink: bool,
+    retry_config: RetryConfig,
+) -> crate::Result<aws_smithy_client::Client> {
+    // Now build the client.
+    let tls_settings = MaybeTlsSettings::tls_client(tls_options)?;
+
+    let connector = if proxy.enabled {
+        let proxy = build_proxy_connector(tls_settings, proxy)?;
+        let hyper_client = aws_smithy_client::hyper_ext::Adapter::builder().build(proxy);
+        aws_smithy_client::erase::DynConnector::new(hyper_client)
+    } else {
+        let tls_connector = build_tls_connector(tls_settings)?;
+        let hyper_client = aws_smithy_client::hyper_ext::Adapter::builder().build(tls_connector);
+        aws_smithy_client::erase::DynConnector::new(hyper_client)
+    };
+
+    let middleware_builder = ServiceBuilder::new()
+        .layer(CaptureRequestSize::new(is_sink, region))
+        .layer(T::default_middleware());
+    let middleware = DynMiddleware::new(middleware_builder);
+
+    let mut client_builder = Builder::new()
+        .connector(connector)
+        .middleware(middleware)
+        .sleep_impl(Some(Arc::new(TokioSleep)));
+    client_builder.set_retry_config(retry_config.into());
+
+    Ok(client_builder.build())
+}
+
 pub async fn create_client<T: ClientBuilder>(
     auth: &AwsAuthentication,
     region: Option<Region>,
@@ -114,31 +148,8 @@ pub async fn create_client<T: ClientBuilder>(
 
     let config = config_builder.build();
 
-    // Now build the client.
-    let tls_settings = MaybeTlsSettings::tls_client(tls_options)?;
-
-    let connector = if proxy.enabled {
-        let proxy = build_proxy_connector(tls_settings, proxy)?;
-        let hyper_client = aws_smithy_client::hyper_ext::Adapter::builder().build(proxy);
-        aws_smithy_client::erase::DynConnector::new(hyper_client)
-    } else {
-        let tls_connector = build_tls_connector(tls_settings)?;
-        let hyper_client = aws_smithy_client::hyper_ext::Adapter::builder().build(tls_connector);
-        aws_smithy_client::erase::DynConnector::new(hyper_client)
-    };
-
-    let middleware_builder = ServiceBuilder::new()
-        .layer(CaptureRequestSize::new(is_sink, region))
-        .layer(T::default_middleware());
-    let middleware = DynMiddleware::new(middleware_builder);
-
-    let mut client_builder = Builder::new()
-        .connector(connector)
-        .middleware(middleware)
-        .sleep_impl(Some(Arc::new(TokioSleep)));
-    client_builder.set_retry_config(retry_config.into());
-
-    let client = client_builder.build();
+    let client = create_smithy_client::<T>(
+        region, proxy, tls_options, is_sink, retry_config).await?;
 
     Ok(T::build(client, &config))
 }

--- a/src/aws/mod.rs
+++ b/src/aws/mod.rs
@@ -89,7 +89,6 @@ pub async fn create_smithy_client<T: ClientBuilder>(
     is_sink: bool,
     retry_config: RetryConfig,
 ) -> crate::Result<aws_smithy_client::Client> {
-    // Now build the client.
     let tls_settings = MaybeTlsSettings::tls_client(tls_options)?;
 
     let connector = if proxy.enabled {
@@ -116,6 +115,18 @@ pub async fn create_smithy_client<T: ClientBuilder>(
     Ok(client_builder.build())
 }
 
+pub async fn resolve_region(region: Option<Region>) -> crate::Result<Region> {
+    match region {
+        Some(region) => Ok(region),
+        None => aws_config::default_provider::region::default_provider()
+            .region()
+            .await
+            .ok_or_else(|| {
+                "Could not determine region from Vector configuration or default providers".into()
+            }),
+    }
+}
+
 pub async fn create_client<T: ClientBuilder>(
     auth: &AwsAuthentication,
     region: Option<Region>,
@@ -128,13 +139,7 @@ pub async fn create_client<T: ClientBuilder>(
 
     // The default credentials chains will look for a region if not given but we'd like to
     // error up front if later SDK calls will fail due to lack of region configuration
-    let region = match region {
-        Some(region) => Ok(region),
-        None => aws_config::default_provider::region::default_provider()
-            .region()
-            .await
-            .ok_or("Could not determine region from Vector configuration or default providers"),
-    }?;
+    let region = resolve_region(region).await?;
 
     // Build the configuration first.
     let mut config_builder = SdkConfig::builder()

--- a/src/aws/mod.rs
+++ b/src/aws/mod.rs
@@ -83,7 +83,7 @@ pub trait ClientBuilder {
 }
 
 pub async fn create_smithy_client<T: ClientBuilder>(
-    region: Region, 
+    region: Region,
     proxy: &ProxyConfig,
     tls_options: &Option<TlsConfig>,
     is_sink: bool,
@@ -148,8 +148,8 @@ pub async fn create_client<T: ClientBuilder>(
 
     let config = config_builder.build();
 
-    let client = create_smithy_client::<T>(
-        region, proxy, tls_options, is_sink, retry_config).await?;
+    let client =
+        create_smithy_client::<T>(region, proxy, tls_options, is_sink, retry_config).await?;
 
     Ok(T::build(client, &config))
 }

--- a/src/sinks/aws_cloudwatch_logs/config.rs
+++ b/src/sinks/aws_cloudwatch_logs/config.rs
@@ -6,7 +6,9 @@ use serde::{Deserialize, Serialize};
 use tower::ServiceBuilder;
 
 use crate::{
-    aws::{create_client, create_smithy_client, AwsAuthentication, ClientBuilder, RegionOrEndpoint},
+    aws::{
+        create_client, create_smithy_client, AwsAuthentication, ClientBuilder, RegionOrEndpoint,
+    },
     codecs::Encoder,
     config::{
         log_schema, AcknowledgementsConfig, GenerateConfig, Input, ProxyConfig, SinkConfig,
@@ -21,9 +23,8 @@ use crate::{
             encoding::{
                 EncodingConfig, EncodingConfigAdapter, StandardEncodings, StandardEncodingsMigrator,
             },
-            http::{RequestConfig},
-            BatchConfig, Compression, ServiceBuilderExt, SinkBatchSettings,
-            TowerRequestConfig,
+            http::RequestConfig,
+            BatchConfig, Compression, ServiceBuilderExt, SinkBatchSettings, TowerRequestConfig,
         },
         Healthcheck, VectorSink,
     },
@@ -90,7 +91,10 @@ impl CloudwatchLogsSinkConfig {
         .await
     }
 
-    pub async fn create_smithy_client(&self, proxy: &ProxyConfig) -> crate::Result<aws_smithy_client::Client> {
+    pub async fn create_smithy_client(
+        &self,
+        proxy: &ProxyConfig,
+    ) -> crate::Result<aws_smithy_client::Client> {
         let region = match self.region.region() {
             Some(region) => Ok(region),
             None => aws_config::default_provider::region::default_provider()
@@ -111,10 +115,13 @@ impl CloudwatchLogsSinkConfig {
 
 #[async_trait::async_trait]
 #[typetag::serde(name = "aws_cloudwatch_logs")]
-impl SinkConfig for CloudwatchLogsSinkConfig{
+impl SinkConfig for CloudwatchLogsSinkConfig {
     async fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
         let batcher_settings = self.batch.into_batcher_settings()?;
-        let request_settings = self.request.tower.unwrap_with(&TowerRequestConfig::default());
+        let request_settings = self
+            .request
+            .tower
+            .unwrap_with(&TowerRequestConfig::default());
         let client = self.create_client(cx.proxy()).await?;
         let smithy_client = self.create_smithy_client(cx.proxy()).await?;
         let svc = ServiceBuilder::new()

--- a/src/sinks/aws_cloudwatch_logs/config.rs
+++ b/src/sinks/aws_cloudwatch_logs/config.rs
@@ -1,4 +1,3 @@
-use aws_config::meta::region::ProvideRegion;
 use aws_sdk_cloudwatchlogs::Client as CloudwatchLogsClient;
 use aws_smithy_types::retry::RetryConfig;
 use futures::FutureExt;
@@ -7,7 +6,8 @@ use tower::ServiceBuilder;
 
 use crate::{
     aws::{
-        create_client, create_smithy_client, AwsAuthentication, ClientBuilder, RegionOrEndpoint,
+        create_client, create_smithy_client, resolve_region, AwsAuthentication, ClientBuilder,
+        RegionOrEndpoint,
     },
     codecs::Encoder,
     config::{
@@ -95,13 +95,7 @@ impl CloudwatchLogsSinkConfig {
         &self,
         proxy: &ProxyConfig,
     ) -> crate::Result<aws_smithy_client::Client> {
-        let region = match self.region.region() {
-            Some(region) => Ok(region),
-            None => aws_config::default_provider::region::default_provider()
-                .region()
-                .await
-                .ok_or("Could not determine region from Vector configuration or default providers"),
-        }?;
+        let region = resolve_region(self.region.region()).await?;
         create_smithy_client::<CloudwatchLogsClientBuilder>(
             region,
             proxy,
@@ -129,7 +123,7 @@ impl SinkConfig for CloudwatchLogsSinkConfig {
             .service(CloudwatchLogsPartitionSvc::new(
                 self.clone(),
                 client.clone(),
-                std::sync::Arc::new(smithy_client).clone(),
+                std::sync::Arc::new(smithy_client),
             ));
         let transformer = self.encoding.transformer();
         let serializer = self.encoding.clone().encoding();

--- a/src/sinks/aws_cloudwatch_logs/config.rs
+++ b/src/sinks/aws_cloudwatch_logs/config.rs
@@ -1,10 +1,12 @@
+use aws_config::meta::region::ProvideRegion;
 use aws_sdk_cloudwatchlogs::Client as CloudwatchLogsClient;
+use aws_smithy_types::retry::RetryConfig;
 use futures::FutureExt;
 use serde::{Deserialize, Serialize};
 use tower::ServiceBuilder;
 
 use crate::{
-    aws::{create_client, AwsAuthentication, ClientBuilder, RegionOrEndpoint},
+    aws::{create_client, create_smithy_client, AwsAuthentication, ClientBuilder, RegionOrEndpoint},
     codecs::Encoder,
     config::{
         log_schema, AcknowledgementsConfig, GenerateConfig, Input, ProxyConfig, SinkConfig,
@@ -19,7 +21,9 @@ use crate::{
             encoding::{
                 EncodingConfig, EncodingConfigAdapter, StandardEncodings, StandardEncodingsMigrator,
             },
-            BatchConfig, Compression, ServiceBuilderExt, SinkBatchSettings, TowerRequestConfig,
+            http::{RequestConfig},
+            BatchConfig, Compression, ServiceBuilderExt, SinkBatchSettings,
+            TowerRequestConfig,
         },
         Healthcheck, VectorSink,
     },
@@ -59,7 +63,7 @@ pub struct CloudwatchLogsSinkConfig {
     #[serde(default)]
     pub batch: BatchConfig<CloudwatchLogsDefaultBatchSettings>,
     #[serde(default)]
-    pub request: TowerRequestConfig,
+    pub request: RequestConfig,
     pub tls: Option<TlsConfig>,
     // Deprecated name. Moved to auth.
     pub assume_role: Option<String>,
@@ -85,20 +89,40 @@ impl CloudwatchLogsSinkConfig {
         )
         .await
     }
+
+    pub async fn create_smithy_client(&self, proxy: &ProxyConfig) -> crate::Result<aws_smithy_client::Client> {
+        let region = match self.region.region() {
+            Some(region) => Ok(region),
+            None => aws_config::default_provider::region::default_provider()
+                .region()
+                .await
+                .ok_or("Could not determine region from Vector configuration or default providers"),
+        }?;
+        create_smithy_client::<CloudwatchLogsClientBuilder>(
+            region,
+            proxy,
+            &self.tls,
+            true,
+            RetryConfig::disabled(),
+        )
+        .await
+    }
 }
 
 #[async_trait::async_trait]
 #[typetag::serde(name = "aws_cloudwatch_logs")]
-impl SinkConfig for CloudwatchLogsSinkConfig {
+impl SinkConfig for CloudwatchLogsSinkConfig{
     async fn build(&self, cx: SinkContext) -> crate::Result<(VectorSink, Healthcheck)> {
         let batcher_settings = self.batch.into_batcher_settings()?;
-        let request_settings = self.request.unwrap_with(&TowerRequestConfig::default());
+        let request_settings = self.request.tower.unwrap_with(&TowerRequestConfig::default());
         let client = self.create_client(cx.proxy()).await?;
+        let smithy_client = self.create_smithy_client(cx.proxy()).await?;
         let svc = ServiceBuilder::new()
             .settings(request_settings, CloudwatchRetryLogic::new())
             .service(CloudwatchLogsPartitionSvc::new(
                 self.clone(),
                 client.clone(),
+                std::sync::Arc::new(smithy_client).clone(),
             ));
         let transformer = self.encoding.transformer();
         let serializer = self.encoding.clone().encoding();

--- a/src/sinks/aws_cloudwatch_logs/request.rs
+++ b/src/sinks/aws_cloudwatch_logs/request.rs
@@ -19,7 +19,7 @@ use futures::{future::BoxFuture, ready, FutureExt};
 use indexmap::IndexMap;
 use tokio::sync::oneshot;
 
-use crate::sinks::aws_cloudwatch_logs::service::CloudwatchError;
+use crate::sinks::aws_cloudwatch_logs::service::{CloudwatchError, SmithyClient};
 
 pub struct CloudwatchFuture {
     client: Client,
@@ -36,12 +36,7 @@ struct Client {
     // client cannot set headers
     //
     // https://github.com/awslabs/aws-sdk-rust/issues/537
-    smithy_client: std::sync::Arc<
-        aws_smithy_client::Client<
-            aws_smithy_client::erase::DynConnector,
-            aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
-        >,
-    >,
+    smithy_client: SmithyClient,
     stream_name: String,
     group_name: String,
     headers: IndexMap<String, String>,
@@ -61,12 +56,7 @@ impl CloudwatchFuture {
     #[allow(clippy::too_many_arguments)]
     pub(super) fn new(
         client: CloudwatchLogsClient,
-        smithy_client: std::sync::Arc<
-            aws_smithy_client::Client<
-                aws_smithy_client::erase::DynConnector,
-                aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
-            >,
-        >,
+        smithy_client: SmithyClient,
         headers: IndexMap<String, String>,
         stream_name: String,
         group_name: String,
@@ -236,7 +226,7 @@ impl Client {
         sequence_token: Option<String>,
         log_events: Vec<InputLogEvent>,
     ) -> ClientResult<PutLogEventsOutput, PutLogEventsError> {
-        let client = self.smithy_client.clone();
+        let client = std::sync::Arc::clone(&self.smithy_client);
         let cw_client = self.client.clone();
         let group_name = self.group_name.clone();
         let stream_name = self.stream_name.clone();

--- a/src/sinks/aws_cloudwatch_logs/request.rs
+++ b/src/sinks/aws_cloudwatch_logs/request.rs
@@ -8,11 +8,15 @@ use aws_sdk_cloudwatchlogs::error::{
     CreateLogGroupError, CreateLogGroupErrorKind, CreateLogStreamError, CreateLogStreamErrorKind,
     DescribeLogStreamsError, DescribeLogStreamsErrorKind, PutLogEventsError,
 };
+use aws_sdk_cloudwatchlogs::operation::PutLogEvents;
+
 use aws_sdk_cloudwatchlogs::model::InputLogEvent;
 use aws_sdk_cloudwatchlogs::output::{DescribeLogStreamsOutput, PutLogEventsOutput};
 use aws_sdk_cloudwatchlogs::types::SdkError;
 use aws_sdk_cloudwatchlogs::Client as CloudwatchLogsClient;
+use aws_smithy_http::operation::{Operation, Request};
 use futures::{future::BoxFuture, ready, FutureExt};
+use indexmap::IndexMap;
 use tokio::sync::oneshot;
 
 use crate::sinks::aws_cloudwatch_logs::service::CloudwatchError;
@@ -28,8 +32,11 @@ pub struct CloudwatchFuture {
 
 struct Client {
     client: CloudwatchLogsClient,
+    smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
+    aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
     stream_name: String,
     group_name: String,
+    headers: IndexMap<String, String>,
 }
 
 type ClientResult<T, E> = BoxFuture<'static, Result<T, SdkError<E>>>;
@@ -46,6 +53,9 @@ impl CloudwatchFuture {
     #[allow(clippy::too_many_arguments)]
     pub(super) fn new(
         client: CloudwatchLogsClient,
+        smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
+        aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
+        headers: IndexMap<String, String>,
         stream_name: String,
         group_name: String,
         create_missing_group: bool,
@@ -56,8 +66,10 @@ impl CloudwatchFuture {
     ) -> Self {
         let client = Client {
             client,
+            smithy_client,
             stream_name,
             group_name,
+            headers,
         };
 
         let state = if let Some(token) = token {
@@ -212,18 +224,42 @@ impl Client {
         sequence_token: Option<String>,
         log_events: Vec<InputLogEvent>,
     ) -> ClientResult<PutLogEventsOutput, PutLogEventsError> {
-        let client = self.client.clone();
+        let client = self.smithy_client.clone();
+        let cw_client = self.client.clone();
         let group_name = self.group_name.clone();
         let stream_name = self.stream_name.clone();
+        let headers = self.headers.clone();
         Box::pin(async move {
-            client
-                .put_log_events()
-                .set_log_events(Some(log_events))
-                .set_sequence_token(sequence_token)
-                .log_group_name(group_name)
-                .log_stream_name(stream_name)
-                .send()
-                .await
+            // #12760 this is a relatively convoluted way of changing the headers of a request
+            // about to be sent. https://github.com/awslabs/aws-sdk-rust/issues/537 should
+            // eventually make this better.
+            let op = PutLogEvents::builder()
+            .set_log_events(Some(log_events))
+            .set_sequence_token(sequence_token)
+            .log_group_name(group_name)
+            .log_stream_name(stream_name)
+            .build()
+            .map_err(|err| aws_smithy_http::result::SdkError::ConstructionFailure(err.into()))?
+            .make_operation(cw_client.conf())
+            .await
+            .map_err(|err| {
+                aws_smithy_http::result::SdkError::ConstructionFailure(err.into())
+            })?;
+            
+            let (req, parts) = op.into_request_response();
+            let (mut body, props) = req.into_parts();
+            for(header, value) in headers.iter() {
+                let owned_header = header.clone();
+                let owned_value = value.clone();
+                body.headers_mut().insert(
+                    http::header::HeaderName::from_bytes(owned_header.as_bytes()).map_err(|err| {
+                        aws_smithy_http::result::SdkError::ConstructionFailure(err.into())
+                    })?,
+                    http::HeaderValue::from_str(owned_value.as_str()).map_err(|err| {
+                        aws_smithy_http::result::SdkError::ConstructionFailure(err.into())
+                    })?);
+            }
+            client.call(Operation::from_parts(Request::from_parts(body, props), parts)).await
         })
     }
 

--- a/src/sinks/aws_cloudwatch_logs/request.rs
+++ b/src/sinks/aws_cloudwatch_logs/request.rs
@@ -32,6 +32,10 @@ pub struct CloudwatchFuture {
 
 struct Client {
     client: CloudwatchLogsClient,
+    // we store a separate smithy_client to set request headers for PutLogEvents since the regular
+    // client cannot set headers
+    //
+    // https://github.com/awslabs/aws-sdk-rust/issues/537
     smithy_client: std::sync::Arc<
         aws_smithy_client::Client<
             aws_smithy_client::erase::DynConnector,

--- a/src/sinks/aws_cloudwatch_logs/request.rs
+++ b/src/sinks/aws_cloudwatch_logs/request.rs
@@ -32,8 +32,12 @@ pub struct CloudwatchFuture {
 
 struct Client {
     client: CloudwatchLogsClient,
-    smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
-    aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
+    smithy_client: std::sync::Arc<
+        aws_smithy_client::Client<
+            aws_smithy_client::erase::DynConnector,
+            aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
+        >,
+    >,
     stream_name: String,
     group_name: String,
     headers: IndexMap<String, String>,
@@ -53,8 +57,12 @@ impl CloudwatchFuture {
     #[allow(clippy::too_many_arguments)]
     pub(super) fn new(
         client: CloudwatchLogsClient,
-        smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
-        aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
+        smithy_client: std::sync::Arc<
+            aws_smithy_client::Client<
+                aws_smithy_client::erase::DynConnector,
+                aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
+            >,
+        >,
         headers: IndexMap<String, String>,
         stream_name: String,
         group_name: String,
@@ -234,32 +242,38 @@ impl Client {
             // about to be sent. https://github.com/awslabs/aws-sdk-rust/issues/537 should
             // eventually make this better.
             let op = PutLogEvents::builder()
-            .set_log_events(Some(log_events))
-            .set_sequence_token(sequence_token)
-            .log_group_name(group_name)
-            .log_stream_name(stream_name)
-            .build()
-            .map_err(|err| aws_smithy_http::result::SdkError::ConstructionFailure(err.into()))?
-            .make_operation(cw_client.conf())
-            .await
-            .map_err(|err| {
-                aws_smithy_http::result::SdkError::ConstructionFailure(err.into())
-            })?;
-            
+                .set_log_events(Some(log_events))
+                .set_sequence_token(sequence_token)
+                .log_group_name(group_name)
+                .log_stream_name(stream_name)
+                .build()
+                .map_err(|err| aws_smithy_http::result::SdkError::ConstructionFailure(err.into()))?
+                .make_operation(cw_client.conf())
+                .await
+                .map_err(|err| {
+                    aws_smithy_http::result::SdkError::ConstructionFailure(err.into())
+                })?;
+
             let (req, parts) = op.into_request_response();
             let (mut body, props) = req.into_parts();
-            for(header, value) in headers.iter() {
+            for (header, value) in headers.iter() {
                 let owned_header = header.clone();
                 let owned_value = value.clone();
                 body.headers_mut().insert(
-                    http::header::HeaderName::from_bytes(owned_header.as_bytes()).map_err(|err| {
-                        aws_smithy_http::result::SdkError::ConstructionFailure(err.into())
-                    })?,
+                    http::header::HeaderName::from_bytes(owned_header.as_bytes()).map_err(
+                        |err| aws_smithy_http::result::SdkError::ConstructionFailure(err.into()),
+                    )?,
                     http::HeaderValue::from_str(owned_value.as_str()).map_err(|err| {
                         aws_smithy_http::result::SdkError::ConstructionFailure(err.into())
-                    })?);
+                    })?,
+                );
             }
-            client.call(Operation::from_parts(Request::from_parts(body, props), parts)).await
+            client
+                .call(Operation::from_parts(
+                    Request::from_parts(body, props),
+                    parts,
+                ))
+                .await
         })
     }
 

--- a/src/sinks/aws_cloudwatch_logs/service.rs
+++ b/src/sinks/aws_cloudwatch_logs/service.rs
@@ -123,6 +123,10 @@ impl CloudwatchLogsPartitionSvc {
     pub fn new(
         config: CloudwatchLogsSinkConfig,
         client: CloudwatchLogsClient,
+        // we store a separate smithy_client to set request headers for PutLogEvents since the regular
+        // client cannot set headers
+        //
+        // https://github.com/awslabs/aws-sdk-rust/issues/537
         smithy_client: std::sync::Arc<
             aws_smithy_client::Client<
                 aws_smithy_client::erase::DynConnector,

--- a/src/sinks/aws_cloudwatch_logs/service.rs
+++ b/src/sinks/aws_cloudwatch_logs/service.rs
@@ -13,6 +13,7 @@ use aws_sdk_cloudwatchlogs::Client as CloudwatchLogsClient;
 use chrono::Duration;
 use futures::{future::BoxFuture, ready, FutureExt};
 use futures_util::TryFutureExt;
+use indexmap::IndexMap;
 use tokio::sync::oneshot;
 use tower::{
     buffer::Buffer,
@@ -119,14 +120,17 @@ impl DriverResponse for CloudwatchResponse {
 }
 
 impl CloudwatchLogsPartitionSvc {
-    pub fn new(config: CloudwatchLogsSinkConfig, client: CloudwatchLogsClient) -> Self {
-        let request_settings = config.request.unwrap_with(&TowerRequestConfig::default());
+    pub fn new(config: CloudwatchLogsSinkConfig, client: CloudwatchLogsClient,
+    smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
+    aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>) -> Self {
+        let request_settings = config.request.tower.unwrap_with(&TowerRequestConfig::default());
 
         Self {
             config,
             clients: HashMap::new(),
             request_settings,
             client,
+            smithy_client,
         }
     }
 }
@@ -172,9 +176,10 @@ impl Service<BatchCloudwatchRequest> for CloudwatchLogsPartitionSvc {
                 .buffer(1)
                 .timeout(self.request_settings.timeout)
                 .service(CloudwatchLogsSvc::new(
-                    &self.config,
+                    self.config.clone(),
                     &key,
                     self.client.clone(),
+                    self.smithy_client.clone(),
                 ));
 
             self.clients.insert(key, svc.clone());
@@ -193,9 +198,11 @@ impl Service<BatchCloudwatchRequest> for CloudwatchLogsPartitionSvc {
 
 impl CloudwatchLogsSvc {
     pub fn new(
-        config: &CloudwatchLogsSinkConfig,
+        config: CloudwatchLogsSinkConfig,
         key: &CloudwatchKey,
         client: CloudwatchLogsClient,
+        smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
+        aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
     ) -> Self {
         let group_name = key.group.clone();
         let stream_name = key.stream.clone();
@@ -204,7 +211,9 @@ impl CloudwatchLogsSvc {
         let create_missing_stream = config.create_missing_stream.unwrap_or(true);
 
         CloudwatchLogsSvc {
+            headers: config.request.headers,
             client,
+            smithy_client,
             stream_name,
             group_name,
             create_missing_group,
@@ -284,6 +293,8 @@ impl Service<Vec<InputLogEvent>> for CloudwatchLogsSvc {
 
             request::CloudwatchFuture::new(
                 self.client.clone(),
+                self.smithy_client.clone(),
+                self.headers.clone(),
                 self.stream_name.clone(),
                 self.group_name.clone(),
                 self.create_missing_group,
@@ -300,6 +311,9 @@ impl Service<Vec<InputLogEvent>> for CloudwatchLogsSvc {
 
 pub struct CloudwatchLogsSvc {
     client: CloudwatchLogsClient,
+    smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
+    aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
+    headers: IndexMap<String, String>,
     stream_name: String,
     group_name: String,
     create_missing_group: bool,
@@ -320,4 +334,6 @@ pub struct CloudwatchLogsPartitionSvc {
     clients: HashMap<CloudwatchKey, Svc>,
     request_settings: TowerRequestSettings,
     client: CloudwatchLogsClient,
+    smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
+    aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
 }

--- a/src/sinks/aws_cloudwatch_logs/service.rs
+++ b/src/sinks/aws_cloudwatch_logs/service.rs
@@ -120,10 +120,20 @@ impl DriverResponse for CloudwatchResponse {
 }
 
 impl CloudwatchLogsPartitionSvc {
-    pub fn new(config: CloudwatchLogsSinkConfig, client: CloudwatchLogsClient,
-    smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
-    aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>) -> Self {
-        let request_settings = config.request.tower.unwrap_with(&TowerRequestConfig::default());
+    pub fn new(
+        config: CloudwatchLogsSinkConfig,
+        client: CloudwatchLogsClient,
+        smithy_client: std::sync::Arc<
+            aws_smithy_client::Client<
+                aws_smithy_client::erase::DynConnector,
+                aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
+            >,
+        >,
+    ) -> Self {
+        let request_settings = config
+            .request
+            .tower
+            .unwrap_with(&TowerRequestConfig::default());
 
         Self {
             config,
@@ -201,8 +211,12 @@ impl CloudwatchLogsSvc {
         config: CloudwatchLogsSinkConfig,
         key: &CloudwatchKey,
         client: CloudwatchLogsClient,
-        smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
-        aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
+        smithy_client: std::sync::Arc<
+            aws_smithy_client::Client<
+                aws_smithy_client::erase::DynConnector,
+                aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
+            >,
+        >,
     ) -> Self {
         let group_name = key.group.clone();
         let stream_name = key.stream.clone();
@@ -311,8 +325,12 @@ impl Service<Vec<InputLogEvent>> for CloudwatchLogsSvc {
 
 pub struct CloudwatchLogsSvc {
     client: CloudwatchLogsClient,
-    smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
-    aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
+    smithy_client: std::sync::Arc<
+        aws_smithy_client::Client<
+            aws_smithy_client::erase::DynConnector,
+            aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
+        >,
+    >,
     headers: IndexMap<String, String>,
     stream_name: String,
     group_name: String,
@@ -334,6 +352,10 @@ pub struct CloudwatchLogsPartitionSvc {
     clients: HashMap<CloudwatchKey, Svc>,
     request_settings: TowerRequestSettings,
     client: CloudwatchLogsClient,
-    smithy_client: std::sync::Arc<aws_smithy_client::Client<aws_smithy_client::erase::DynConnector,
-    aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>>>,
+    smithy_client: std::sync::Arc<
+        aws_smithy_client::Client<
+            aws_smithy_client::erase::DynConnector,
+            aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
+        >,
+    >,
 }

--- a/src/sinks/aws_cloudwatch_logs/service.rs
+++ b/src/sinks/aws_cloudwatch_logs/service.rs
@@ -50,6 +50,13 @@ type Svc = Buffer<
     Vec<InputLogEvent>,
 >;
 
+pub type SmithyClient = std::sync::Arc<
+    aws_smithy_client::Client<
+        aws_smithy_client::erase::DynConnector,
+        aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
+    >,
+>;
+
 #[derive(Debug)]
 pub enum CloudwatchError {
     Put(SdkError<PutLogEventsError>),
@@ -127,12 +134,7 @@ impl CloudwatchLogsPartitionSvc {
         // client cannot set headers
         //
         // https://github.com/awslabs/aws-sdk-rust/issues/537
-        smithy_client: std::sync::Arc<
-            aws_smithy_client::Client<
-                aws_smithy_client::erase::DynConnector,
-                aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
-            >,
-        >,
+        smithy_client: SmithyClient,
     ) -> Self {
         let request_settings = config
             .request
@@ -193,7 +195,7 @@ impl Service<BatchCloudwatchRequest> for CloudwatchLogsPartitionSvc {
                     self.config.clone(),
                     &key,
                     self.client.clone(),
-                    self.smithy_client.clone(),
+                    std::sync::Arc::clone(&self.smithy_client),
                 ));
 
             self.clients.insert(key, svc.clone());
@@ -215,12 +217,7 @@ impl CloudwatchLogsSvc {
         config: CloudwatchLogsSinkConfig,
         key: &CloudwatchKey,
         client: CloudwatchLogsClient,
-        smithy_client: std::sync::Arc<
-            aws_smithy_client::Client<
-                aws_smithy_client::erase::DynConnector,
-                aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
-            >,
-        >,
+        smithy_client: SmithyClient,
     ) -> Self {
         let group_name = key.group.clone();
         let stream_name = key.stream.clone();
@@ -311,7 +308,7 @@ impl Service<Vec<InputLogEvent>> for CloudwatchLogsSvc {
 
             request::CloudwatchFuture::new(
                 self.client.clone(),
-                self.smithy_client.clone(),
+                std::sync::Arc::clone(&self.smithy_client),
                 self.headers.clone(),
                 self.stream_name.clone(),
                 self.group_name.clone(),
@@ -329,12 +326,7 @@ impl Service<Vec<InputLogEvent>> for CloudwatchLogsSvc {
 
 pub struct CloudwatchLogsSvc {
     client: CloudwatchLogsClient,
-    smithy_client: std::sync::Arc<
-        aws_smithy_client::Client<
-            aws_smithy_client::erase::DynConnector,
-            aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
-        >,
-    >,
+    smithy_client: SmithyClient,
     headers: IndexMap<String, String>,
     stream_name: String,
     group_name: String,
@@ -356,10 +348,5 @@ pub struct CloudwatchLogsPartitionSvc {
     clients: HashMap<CloudwatchKey, Svc>,
     request_settings: TowerRequestSettings,
     client: CloudwatchLogsClient,
-    smithy_client: std::sync::Arc<
-        aws_smithy_client::Client<
-            aws_smithy_client::erase::DynConnector,
-            aws_smithy_client::erase::DynMiddleware<aws_smithy_client::erase::DynConnector>,
-        >,
-    >,
+    smithy_client: SmithyClient,
 }

--- a/website/cue/reference/components/sinks/aws_cloudwatch_logs.cue
+++ b/website/cue/reference/components/sinks/aws_cloudwatch_logs.cue
@@ -39,7 +39,7 @@ components: sinks: aws_cloudwatch_logs: components._aws & {
 			proxy: enabled: true
 			request: {
 				enabled: true
-				headers: false
+				headers: true
 			}
 			tls: {
 				enabled:                true

